doc/unboxing.md

```diff
@@ -0,0 +1,234 @@
+Unboxing
+========
+
+## Overview
+
+The unboxing algorithm in Flambda 2 aims to remove unnecessary allocations
+without causing bad behaviour (e.g. pushing allocations into loops).
+It currently supports:
+
+- Unboxing along non-looping control flow, within a single function, of the
+  following varieties of values:
+  - Boxed `float`
+  - Boxed `int32`
+  - Boxed `int64`
+  - Boxed `nativeint`
+  - Variants
+  - Closures (need examples).
+
+- Unboxing around loops, save that the corresponding allocations will only be
+  removed if:
+  - the first use of the value being unboxed in the loop can make use of the
+    unboxed form (for example if it is a floating-point arithmetic
+    operation, as opposed to something like a non-inlined function call); and
+  - this same condition also holds for the first use of the value after
+    the loop, on all exit paths from the loop.  (The trick of "`+. 0.`" can
+    be used here; this will not be optimised away by Flambda 2.)
+
+The unboxing code does not handle the untagging of integers.
```

Comment:
I'm not sure about that point, since there is an unboxer for tagged immediate (cf https://github.com/ocaml-flambda/ocaml/pull/394/files#diff-5158694ba54fff5a292c030274b821c39c9613770aa5843afaf70e598c351068R42 ). It is true that most of the untagging of integers comes from other parts of the optimization process, but the unboxing pass should be able to untag an integer across continuation calls.